Tests/ApolloCodegenTests/CodeGenAST/ASTSelectionSetTests.swift

```diff
@@ -0,0 +1,2415 @@
+  }
+
+  /// Example:
+  /// query {
+  ///  rocks {
+  ///    ...AnimalDetails
+  /// }
+  ///
+  /// fragment AnimalDetails on Animal {
+  ///   species
+  /// }
+  /// Expected:
+  /// Children should not include a type case for asAnimal
+  func test__children__initWithNamedFragmentOnUnrelatedType_hasChildTypeCase() {
+    // given
+    let Interface_Animal = GraphQLInterfaceType.mock("Animal")
+    let Object_Rock = GraphQLObjectType.mock("Rock")
+    let animalDetails = CompilationResult.FragmentDefinition.mock("AnimalDetails", type: Interface_Animal)
+
+    let subject = ASTSelectionSet(selectionSet: .mock(
+      parentType: Object_Rock,
+      selections: [
+        .fragmentSpread(animalDetails),
+      ]
+    ), compilationResult: mockCompilationResult)
+
+    // then
+    expect(subject.children.count).to(equal(1))
+    let child = subject.children["Animal"]!
+    expect(child.parent).to(beIdenticalTo(subject))
+    expect(child.type).to(equal(Interface_Animal))
+    expect(child.selections).to(equal([.fragmentSpread(animalDetails)]))
+  }
+
+  // MARK: Children Computation - Union Type
+
+  /// Example:
+  /// query {
+  ///  allAnimals {
+  ///    ... on ClassroomPet {
+  ///      ... on Bird {
+  ///        species
+  ///      }
+  ///    }
+  ///   }
+  /// }
+  func test__children__givenIsUnionType_withNestedTypeCaseOfObjectType_hasChildrenForTypeCase() {
+    // given
+    let Object_Bird = GraphQLObjectType.mock("Bird")
+    let Union_ClassroomPet = GraphQLUnionType.mock("ClassroomPet", types: [Object_Bird])
+
+    let Field_Species: CompilationResult.Selection = .field(.mock("species"))
+
+    let parent = ASTSelectionSet(selectionSet: .mock(
+      parentType: GraphQLInterfaceType.mock("Animal"),
+      selections: [
+        .inlineFragment(.mock(
+          parentType: Union_ClassroomPet,
+          selections: [
+            .inlineFragment(.mock(
+              parentType: Object_Bird,
+              selections: [Field_Species]
+            ))]
+        )),
+      ]
+    ), compilationResult: mockCompilationResult)
+
+    // when
+    let onClassroomPet = parent.children["ClassroomPet"]!
+    let onClassroomPet_onBird = onClassroomPet.children["Bird"]!
+
+    // then
+    expect(onClassroomPet.parent).to(beIdenticalTo(parent))
+    expect(onClassroomPet.type).to(beIdenticalTo(Union_ClassroomPet))
+    expect(onClassroomPet.children.count).to(equal(1))
+
+    expect(onClassroomPet_onBird.parent).to(beIdenticalTo(onClassroomPet))
+    expect(onClassroomPet_onBird.type).to(beIdenticalTo(Object_Bird))
+    expect(onClassroomPet_onBird.selections).to(equal([Field_Species]))
+  }
+
+  // MARK: Children - Type Cases
+
+  /// Example:
+  /// query { // On A
+  ///   A
+  ///   ... on A {
+  ///     B
+  ///   }
+  /// }
+  ///
+  /// Expected:
+  /// Query.Children: { }
+  func test__children__givenInlineFragment_onSameType_mergesTypeCaseIn_doesNotHaveTypeCaseChild() {
+    // given
+    let Object_A = GraphQLObjectType.mock("A")
```

Comment:
Is there a reason you're using abstract types here instead of examples from the schema? Honestly I find the examples from the schema a hell of a lot easier to follow.

Alternately, is it worth breaking these ones that are more abstract out into a separate file so it's clearer where we're testing actual use cases rather than purely conceptual ones?

Reply:
All of these are conceptual use cases really. None of this is using the actual schema. Each unit test creates the schema types it needs. (For example, In some of the tests, WarmBlooded implements Pet, in others it doesn't).

Most of these that just use "A, B, etc" are contrived examples that were difficult to create a real-world example of that made sense with the AnimalKingdom schema. I didn't want to waste a ton of time trying to come up with concepts that made sense for each example. For instance, Pet, WarmBlooded, and ClassroomPet doesn't always make sense for use cases with different Interfaces on the same type that implemented interfaces on other different sub types and selected the same fields.